### PR TITLE
Change Roslyn complation exception search

### PR DESCRIPTION
**Pull request by pakrym**
Make it recursive so as to handle bot first level and wrapped compilation exception.

/cc @troydai @victorhurdugaci @davidfowl 

Address: #2887

### CODE REVIEW

File: misc/ApplicationHostTestProjects/src/SyntaxException/program.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿class Program
+{
+	public static void Main()
+	{
+		System.Console1.WriteLine("Test");
```
Comment (davidfowl) [inline misc/ApplicationHostTestProjects/src/SyntaxException/program.cs]: No tabs please.

File: misc/ApplicationHostTestProjects/src/SyntaxException/project.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+﻿{
+   "frameworks": {
+        "dnx451": { },
+        "dnxcore50": {
+            "dependencies": {
+                "System.Runtime": "4.0.21-beta-*"
+            }
```
Comment (davidfowl) [inline misc/ApplicationHostTestProjects/src/SyntaxException/project.json]: 2 spaces not 4

File: src/Microsoft.Dnx.ApplicationHost/Program.cs [hunk 3]
```diff
@@ -229,19 +229,29 @@ private Task<int> ExecuteMain(DefaultHost host, string applicationName, string[]
...
-            // from version to version. It's safest to catch both types of exceptions
-            catch (Exception ex) when (ex is FileLoadException || ex is FileNotFoundException)
+            catch (Exception ex)
             {
-                if (ex.InnerException is ICompilationException)
+                // Try to find compilation exception recursively to supress its stack
+                var innerException = ex;
+                do
                 {
-                    throw SuppressStackTrace(ex.InnerException);
```
Comment (davidfowl) [inline src/Microsoft.Dnx.ApplicationHost/Program.cs]: When would you need to dive like this?

File: misc/ApplicationHostTestProjects/src/SyntaxException/program.cs (hunk 1 above)
Comment (victorhurdugaci) [inline misc/ApplicationHostTestProjects/src/SyntaxException/program.cs]: Using a type name that doesn't exist is not a syntax error. Either rename the project or add an actual syntax error

File: src/Microsoft.Dnx.ApplicationHost/Program.cs (hunk 3 above)
Comment (pakrym) [inline src/Microsoft.Dnx.ApplicationHost/Program.cs]: The way dnx loads assembly when running on clr\coreclr differs so we get ICompilationException  either wrapped into FileLoadException or first level. So its either loop trough inner exceptions searching for ICompilationException or two if's like:

    if (ex is ICompilationException)
    {
        throw SuppressStackTrace(ex);
    }

    if (ex.InnerException is ICompilationException)
    {
        throw SuppressStackTrace(ex.InnerException);
    }

File: test/Microsoft.Dnx.ApplicationHost.FunctionalTests/AppHostTests.cs [hunk 4]
```diff
@@ -90,5 +90,27 @@ public void RunP2PDifferentFrameworks(DnxSdk sdk)
 
             TestUtils.CleanUpTestDir<AppHostTests>(sdk);
         }
+        
+        [Theory, TraceTest]
+        [MemberData(nameof(DnxSdks))]
+        public void DxtSupressesRoslynExceptionStackTrace(DnxSdk sdk)
```
Comment (victorhurdugaci) [inline test/Microsoft.Dnx.ApplicationHost.FunctionalTests/AppHostTests.cs]: Dxt?

File: test/Microsoft.Dnx.ApplicationHost.FunctionalTests/AppHostTests.cs [hunk 5]
```diff
@@ -90,5 +90,27 @@ public void RunP2PDifferentFrameworks(DnxSdk sdk)
...
+            var project = solution.GetProject("CompilationException");
+
+            sdk.Dnu.Restore(solution.RootPath).EnsureSuccess();
+
+            // Act
+            var result = sdk.Dnx.Execute(project);
+
+            // Assert
+            Assert.Equal(1, result.ExitCode);
+            Assert.DoesNotContain("Exception", result.StandardOutput);
```
Comment (victorhurdugaci) [inline test/Microsoft.Dnx.ApplicationHost.FunctionalTests/AppHostTests.cs]: How about StdErr? Usually error messages end up there

**Comment by pakrym** [comment]
ping @troydai @victorhurdugaci @davidfowl

File: src/Microsoft.Dnx.ApplicationHost/Program.cs [hunk 6]
```diff
@@ -229,19 +229,29 @@ private Task<int> ExecuteMain(DefaultHost host, string applicationName, string[]
...
+                    if (innerException is ICompilationException)
+                    {
+                        throw SuppressStackTrace(innerException);
+                    }
+                    innerException = innerException.InnerException;
```
Comment (troydai) [inline src/Microsoft.Dnx.ApplicationHost/Program.cs]: Could the `innerException == null` here (therefore cause `NullRef`)? There is nothing guard this from happening. Or is there any implicit rules there that the captured the exception must have an inner exception?

**Comment by troydai** [comment]
:shipit:

**Comment by pakrym** [comment]
@JunTaoLuo  @davidfowl  What if user writes console application using Roslyn and he has roslyn exception inside of it. We would suppress it's stack trace if we put try/catch around ExecuteMain.

**Comment by davidfowl** [comment]
ICompilationException is something that we expose. If you use roslyn directly they won't get exceptions for errors. If you throw an exception that implements ICompilationException, you get our default suppression behavior.

**Comment by pakrym** [comment]
Reopened the issue https://github.com/aspnet/dnx/issues/2887

**Pull request merged**